src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -1282,4 +1282,47 @@ BOOST_AUTO_TEST_CASE(SimpleMeanSquaredErrorLayerTest)
   BOOST_REQUIRE_EQUAL(output.n_elem, 1);
 }
 
+/*
+ * Simple test for the BilinearInterpolation layer
+ */
+BOOST_AUTO_TEST_CASE(SimpleBilinearInterpolationLayerTest)
+{
+  arma::mat input, output, unzoomedOutput, expectedOutput;
+  size_t inRowSize = 10;
+  size_t inColSize = 10;
+  size_t outRowSize = 20;
+  size_t outColSize = 20;
+  size_t depth = 3;
+  input.zeros(inRowSize * inColSize *  depth, 1);
+  for (size_t d = 0; d < depth; d++)
+    for (size_t i = 0; i < inRowSize; i++)
+      for (size_t j = 0; j < inColSize; j++)
+      {
+        size_t ptr = d * inRowSize * inColSize + j * inColSize + i;
+        input[ptr] =  1 * (double)i / inRowSize +
+            1 * double(j) / inColSize + 1 * double(d) / depth;
+      }
+
+  expectedOutput.set_size(outRowSize * outColSize * depth, 1);
+  for (size_t d = 0; d < depth; d++)
+    for (size_t i = 0; i < outRowSize; i++)
+      for (size_t j = 0; j < outColSize; j++)
+      {
+        size_t ptr = d * outRowSize * outColSize + j * outColSize + i;
+        expectedOutput[ptr] = 1 * i / (double)outRowSize +
+            1 * j / (double)outColSize + 1 * d / (double)depth;
+      }
+
+  BilinearInterpolation<> layer(inRowSize, inColSize, outRowSize, outColSize,
+                                depth);
+
+  layer.Forward(std::move(input), std::move(output));
+  CheckMatrices(output - expectedOutput, arma::zeros(output.n_rows), 1e-12);
+
+
+  layer.Backward(std::move(output), std::move(output),
+        std::move(unzoomedOutput));
+  CheckMatrices(unzoomedOutput - input, arma::zeros(input.n_rows), 1e-12);
+}
```

Comment:
Do you think you could add a numerical gradient test like, for instance, `GradientLSTMLayerTest`?  These use the `AddGradient()` function, and it should be simple to adapt the existing tests to this layer type.

Reply:
Yes, the ``CheckGradient`` function: https://github.com/mlpack/mlpack/blob/346fd1f819122cdc6b5cf148291dcec05e1dcbce/src/mlpack/tests/ann_layer_test.cpp#L146-L178

Reply:
Oops, you are right, sorry for my error.  I did indeed mean `CheckGradient()`.  It's still a useful test, even if the gradient is the same as the propagated error.

Reply:
My only remaining comment is that I'd still like to see a test for the gradient.  Even though the layer has no Gradient() method, that means the gradient should implicitly be all ones.  So we can adapt the function @zoq mentioned above and check against that.  In my view this is still an important check to add.

Reply:
@rcurtin I did add a test like that, but the issue with that test was that `costPlus` and `costMinus` values returned the same `double` value, and hence the `estGradient` became `0` for each output point. Hence @zoq's comment [here](https://github.com/mlpack/mlpack/pull/1238#discussion_r169102607).

Reply:
Thanks for the link.  Github had hidden those comments.  In that case there are no more comments on this PR from me.  Thank you for working on it!